src/gridfs/upload.rs

```diff
@@ -0,0 +1,147 @@
+        filename: impl AsRef<str>,
+        source: T,
+        options: impl Into<Option<GridFsUploadOptions>>,
+    ) -> Result<ObjectId>
+    where
+        T: AsyncRead + Unpin,
+    {
+        let id = ObjectId::new();
+        self.upload_from_futures_0_3_reader_with_id(id.into(), filename, source, options)
+            .await?;
+        Ok(id)
+    }
+
+    /// Uploads a user file to a GridFS bucket with the given `files_id`. Bytes are read from
+    /// `source` and stored in chunks in the bucket's chunks collection. After all the chunks have
+    /// been uploaded, a corresponding [`FilesCollectionDocument`] is stored in the bucket's files
+    /// collection.
+    pub async fn upload_from_futures_0_3_reader_with_id<T>(
+        &self,
+        files_id: Bson,
+        filename: impl AsRef<str>,
+        mut source: T,
+        options: impl Into<Option<GridFsUploadOptions>>,
+    ) -> Result<()>
+    where
+        T: AsyncRead + Unpin,
+    {
+        let options = options.into();
+
+        self.create_indexes().await?;
+
+        let chunk_size = options
+            .as_ref()
+            .and_then(|opts| opts.chunk_size_bytes)
+            .unwrap_or_else(|| self.chunk_size_bytes());
+        let mut length = 0u64;
+        let mut n = 0;
+
+        loop {
+            let mut buf = vec![0u8; chunk_size as usize];
+            let bytes_read = match source.read(&mut buf).await {
+                Ok(0) => break,
+                Ok(n) => n,
+                Err(error) => {
+                    self.chunks()
+                        .delete_many(doc! { "files_id": &files_id }, None)
+                        .await?;
+                    return Err(ErrorKind::Io(error.into()).into());
+                }
+            };
+
+            buf.truncate(bytes_read);
+            let chunk = Chunk {
+                id: ObjectId::new(),
+                files_id: files_id.clone(),
+                n,
+                data: RawBinaryRef {
+                    subtype: BinarySubtype::Generic,
+                    bytes: &buf,
+                },
+            };
+            self.chunks().insert_one(chunk, None).await?;
+
+            length += bytes_read as u64;
+            n += 1;
+        }
+
+        let file = FilesCollectionDocument {
+            id: files_id,
+            length,
+            chunk_size,
+            upload_date: DateTime::now(),
+            filename: Some(filename.as_ref().to_string()),
+            metadata: options.and_then(|opts| opts.metadata),
+        };
+        self.files().insert_one(file, None).await?;
+
+        Ok(())
+    }
+
+    async fn create_indexes(&self) -> Result<()> {
+        if !self.inner.created_indexes.load(Ordering::SeqCst) {
+            let find_options = FindOneOptions::builder()
+                .selection_criteria(SelectionCriteria::ReadPreference(ReadPreference::Primary))
+                .build();
+            if self.files().find_one(None, find_options).await?.is_none() {
```

Comment:
Good catch, updated. The projection will cause `find_one` to fail on `self.files()` if a doc gets returned because it won't match the `FilesCollectionDocument` schema, so I also added a `clone_with_type::<Document>` call to avoid this.